private/buf/buffetch/ref_parser.go

```diff
@@ -30,50 +30,62 @@ import (
 )
 
 type refParser struct {
-	logger         *zap.Logger
-	fetchRefParser internal.RefParser
+	allowProtoFileRef bool
+	logger            *zap.Logger
+	fetchRefParser    internal.RefParser
 }
 
-func newRefParser(logger *zap.Logger) *refParser {
-	return &refParser{
-		logger: logger.Named("buffetch"),
-		fetchRefParser: internal.NewRefParser(
-			logger,
-			internal.WithRawRefProcessor(processRawRef),
-			internal.WithSingleFormat(formatBin),
-			internal.WithSingleFormat(formatJSON),
-			internal.WithSingleFormat(
-				formatBingz,
-				internal.WithSingleDefaultCompressionType(
-					internal.CompressionTypeGzip,
-				),
-			),
-			internal.WithSingleFormat(
-				formatJSONGZ,
-				internal.WithSingleDefaultCompressionType(
-					internal.CompressionTypeGzip,
-				),
-			),
-			internal.WithArchiveFormat(
-				formatTar,
-				internal.ArchiveTypeTar,
+func newRefParser(logger *zap.Logger, options ...RefParserOption) *refParser {
+	refParser := &refParser{}
+	for _, option := range options {
+		option(refParser)
+	}
+	fetchRefParserOptions := []internal.RefParserOption{
+		internal.WithRawRefProcessor(newRawRefProcessor(&rawRefProcessorConfig{
+			allowProtoFileRef: refParser.allowProtoFileRef,
+		})),
+		internal.WithSingleFormat(formatBin),
+		internal.WithSingleFormat(formatJSON),
+		internal.WithSingleFormat(
+			formatBingz,
+			internal.WithSingleDefaultCompressionType(
+				internal.CompressionTypeGzip,
 			),
-			internal.WithArchiveFormat(
-				formatTargz,
-				internal.ArchiveTypeTar,
-				internal.WithArchiveDefaultCompressionType(
-					internal.CompressionTypeGzip,
-				),
+		),
+		internal.WithSingleFormat(
+			formatJSONGZ,
+			internal.WithSingleDefaultCompressionType(
+				internal.CompressionTypeGzip,
 			),
-			internal.WithArchiveFormat(
-				formatZip,
-				internal.ArchiveTypeZip,
+		),
+		internal.WithArchiveFormat(
+			formatTar,
+			internal.ArchiveTypeTar,
+		),
+		internal.WithArchiveFormat(
+			formatTargz,
+			internal.ArchiveTypeTar,
+			internal.WithArchiveDefaultCompressionType(
+				internal.CompressionTypeGzip,
 			),
-			internal.WithGitFormat(formatGit),
-			internal.WithDirFormat(formatDir),
-			internal.WithModuleFormat(formatMod),
 		),
+		internal.WithArchiveFormat(
+			formatZip,
+			internal.ArchiveTypeZip,
+		),
+		internal.WithGitFormat(formatGit),
+		internal.WithDirFormat(formatDir),
+		internal.WithModuleFormat(formatMod),
 	}
+	if refParser.allowProtoFileRef {
+		fetchRefParserOptions = append(fetchRefParserOptions, internal.WithProtoFileFormat(formatProtoFile))
+	}
+	refParser.logger = logger.Named("buffetch")
+	refParser.fetchRefParser = internal.NewRefParser(
+		logger,
+		fetchRefParserOptions...,
+	)
+	return refParser
```

Comment:
Let's rewrite the `RefParserOption` so that it doesn't mutate the `refParser` type. We can define a new un-exported `refParserOptions` type, similar to what you see elsewhere (e.g. [here](https://github.com/bufbuild/buf/blob/60cedc242b5231d92c37841278e223d6308a6637/private/pkg/app/appproto/response_writer.go#L148)).

With this, we don't have to construct the `refParser` up-front, then mutate it here. We can just apply all the options and construct the `refParser` at the end (similar to what we used to have).

And given that this is an un-exported constructor, the options could be applied in the top-level `NewRefParser` and concrete values could be passed here, e.g.

```
newRefParser(logger *zap.Logger, allowProtoFileFormat bool)
```

Reply:
This was based on a previous comment: https://github.com/bufbuild/buf/pull/608#discussion_r716991316.
The original implementation of this used an intermediary options type.